src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs

```diff
@@ -249,11 +249,14 @@ public bool AreEqual(object x, object y, ref Tolerance tolerance)
 
         private static MethodInfo FirstImplementsIEquatableOfSecond(Type first, Type second)
         {
+            var pair = new KeyValuePair<Type, MethodInfo>();
+
             foreach (var xEquatableArgument in GetEquatableGenericArguments(first))
                 if (xEquatableArgument.Key.IsAssignableFrom(second))
-                    return xEquatableArgument.Value;
+                    if (pair.Key == null || pair.Key.IsAssignableFrom(xEquatableArgument.Key))
```

Comment:
I've merged this PR and created an issue to ensure the order is what we expect. Even if it isn't, I don't think this implementation is any worse than what we already have and on many/most platforms it is better.